src/mlpack/methods/edge_boxes/feature_extraction_impl.hpp

```diff
@@ -0,0 +1,1052 @@
+      Timer::Stop("get_features");
+
+      //randomly sample 70 values each from reg_ftr and ss_ftr.
+      /*
+      CubeType ftr(140, 1000, 13);
+      arma::uvec r = (0, 255, 256);
+      arma::uvec s = (0, 299, 300);
+      arma::uvec rs = r.shuffle();
+      arma::uvec ss = s.shuffle();
+      */
+      //MatType lbl(1000, gSize * gSize);
+      CubeType s(segs.n_rows, segs.n_cols, 1);
+      
+      // have to do this or we can overload the CopyMakeBorder to support MatType.
+      s.slice(0) = segs;
+      CubeType in_segs;
+      this->CopyMakeBorder(s, gRad, gRad, gRad,\
+                            gRad, in_segs);
+
+      for(size_t i = 0; i < loc.n_rows; ++i)
+      {
+        size_t x = loc(i, 0); size_t y = loc(i, 1);
+        //std::cout << "x, y = " << x << " " << y << std::endl;
+        lbls.row(i) = arma::vectorise(in_segs.slice(0)\
+                    .submat((x + gRad) - gRad, (y + gRad) - gRad,\
+                     (x + gRad) + gRad - 1, (y + gRad) + gRad - 1)).t();
+      }
+    }
+    arma::vec DiscreteLabels;
+    size_t x = Discretize(lbls, 2, 256, DiscreteLabels);
+  }
+}
+
+template<typename MatType, typename CubeType>
+size_t StructuredForests<MatType, CubeType>::
+IndexMin(arma::vec& k)
+{
+  double s = k(0); size_t ind = 0;
+  for (size_t i = 1; i < k.n_elem; ++i)
+  {
+    if (k(i) < s)
+    {
+      s = k(i);
+      ind = i;
+    }
+  }
+  return ind;
+}
+// returns the index of the most representative label, and discretizes structured
+// label to discreet classes in matrix subLbls. (this is a vector if nClass = 2)
+template<typename MatType, typename CubeType>
+size_t StructuredForests<MatType, CubeType>::
+Discretize(const MatType& labels, const size_t nClass,\
+           const size_t nSample, arma::vec& DiscreteLabels)
+{
+  // Map labels to discrete class labels.
+  // lbls : 20000 * 256.
+  // nSample: number of samples for clustering structured labels 256
+  // nClass: number of classes (clusters) for binary splits. 2
+  Timer::Start("other_discretize");
+
+  arma::uvec lis1(nSample);  
+  for (size_t i = 0; i < lis1.n_elem; ++i)
+    lis1(i) = i;
+  
+  MatType zs(labels.n_rows, nSample);
+  // no. of principal components to keep.
+  size_t dim = std::min( 5, std::min( (int)nSample,\
+              (int)std::floor( std::log2( (int)nClass ) ) ) );
+  DiscreteLabels = arma::zeros(labels.n_rows, dim);
+  for (size_t j = 0; j < zs.n_cols; ++j)
+  {
+    arma::uvec z1 = arma::shuffle(lis1);
+    arma::uvec z2 = arma::shuffle(lis1);
+    for (size_t i = 0; i < zs.n_rows; ++i)
+      zs(i, j) = (labels(i, z1(j)) == labels(i, z2(j))) ? 1 : 0;
+  }
+  for (size_t i = 0; i < zs.n_cols; ++i)
+    zs.row(i) -= arma::mean(zs, 0); // calculate mean about rows. n_rows = 20000.
+  size_t ind = 0;
+  arma::uvec k = arma::find(zs > 0);
+  if ( k.n_elem == 0)
+  {
+    DiscreteLabels.ones();
+  }
+  else
+  {
+    //find most representative label (closest to mean)
+    arma::vec k = arma::sum(arma::abs(zs), 0);
```

Comment:
As pointed out here: http://arma.sourceforge.net/docs.html#sum ``arma::sum(..., 0)`` returns a ``rowvec``, arma::vec is a ``colvec``.